lib/libconfig.c

```diff
@@ -211,53 +211,65 @@ EXPORTED int config_parseduration(const char *str, int defunit, int *out_duratio
     p = copy;
     if (*p == '-') {
         neg = 1;
-        p++;
+        if (!*++p) {
+            r = -1;
+            goto done;
+        }
     }
-    for (; *p; p++) {
-        if (cyrus_isdigit(*p)) {
-            accum *= 10;
-            accum += (*p - '0');
-            sawdigit = 1;
+
+    while (*p) {
+
+        if (!isdigit(*p)) {
+            r = -1;
+            goto done;
         }
-        else {
-            if (!sawdigit) {
-                syslog(LOG_DEBUG, "%s: no digit before '%c' in '%s'",
-                                  __func__, *p, str);
-                r = -1;
-                goto done;
-            }
-            sawdigit = 0;
-            switch (*p) {
-            case 'd':
-                accum *= 24;
-                /* fall through */
-            case 'h':
-                accum *= 60;
-                /* fall through */
-            case 'm':
-                accum *= 60;
-                /* fall through */
-            case 's':
-                duration += accum;
-                accum = 0;
-                break;
-            default:
-                syslog(LOG_DEBUG, "%s: bad unit '%c' in %s",
-                                  __func__, *p, str);
+
+        uint64_t val = strtoul(p, &end, 10);
+        if (p == end) {
+            r = -1;
+            goto done;
+        }
+
+        switch (*end) {
+        case 'd':
+            val *= 24;
+            /* fall through */
+        case 'h':
+            val *= 60;
+            /* fall through */
+        case 'm':
+            val *= 60;
+            /* fall through */
+        case 's':
+            if (duration + val > INT32_MAX) {
```

Comment:
I can see that you used a similar implementation to cyr_expire's `parse_duration()`, letting `strtoul()` handle the digit parsing instead of accumulating them one at a time.  But this approach only works here because the type being used is `int`, which is smaller than `uint64_t`.  If we did the same thing in `config_parsebytesize()`, which computes an `int64_t` value, the overflow check would itself overflow -- so we couldn't use the same strategy in that one, and I think keeping the strategies similar is important for maintainability.

I was anticipating implementing these fixes similar to `getint32()` in imap/imapparse.c -- which each time it sees a new digit, it checks the current value, and if it's already too big to accept an additional digit, then it bails out.  For `config_parseduration()` and `config_parsebytesize()` it would be a little more complicated than that, because those functions don't only multiply by ten.  But the point is it's important to detect the potential overflow _before_ it happens, because we can't necessarily detect it after.